### PR TITLE
feat: wrapping up student login (adding date authentication & showing teacher notes & removing extraneous fields)

### PR DESCRIPTION
**Pull request by joyce-shi**
## Notion ticket link

[Update Test Session model to support Distributing Assessments](https://www.notion.so/uwblueprintexecs/Update-Test-Session-model-to-support-distributing-assessments-19cb55d738ed43e092cd48c1b4980675?pvs=4)
[Update backend to support Student Authentication by Date](https://www.notion.so/uwblueprintexecs/Update-backend-to-support-Student-Authentication-by-date-f85dd5665e9c44cba169f802edf154db?pvs=4)
[Assessment Summary page with No Teacher Notes](https://www.notion.so/uwblueprintexecs/Assessment-Summary-page-with-No-Teacher-Notes-7eb8e593157a41b78c59c339acd311a7?pvs=4)
[Populate start date and test session notes dynamically in Assessment Summary page](https://www.notion.so/uwblueprintexecs/Populate-start-date-and-test-session-notes-dynamically-in-Assessment-Summary-page-98991998d29241cf947de911097b3cfe?pvs=4)
[Save class name to Student Context](https://www.notion.so/uwblueprintexecs/Store-className-in-Student-Context-03bf1be7e8514e16be5cfc4541140726?pvs=4)

## Implementation description
* added `startDate` and `endDate` fields to test session model
* added optional `notes` field to test session model
* removed `startTime` and `gradeLevel` fields from test session model
* updated `createTestSession` to include `startDate` and `endDate` fields (& checks that they are valid)
* updated `getTestSessionByAccessCode` to include filtering by `startDate` and `endDate`
* unit testing for added backend functionality
* storing `className` in `StudentContext` (this is for the `StudentDashboardHeader` that @JennaXiao is working on)
* added fetching for `AssessmentSummary` page with `startDate` and `notes`
* updated `AssessmentSummary` page to handle no notes case (there is a design for this)

Note: should we force `accessCode` to be 6-digits? or is the frontend check sufficient

## What should reviewers focus on?
* updated UI features
* updated backend functionality works as expected (`getTestSessionByAccessCode`)
* best practices
* that i didn't break anything in the merge since my branch was pretty far behind and i updated a lot of the same files that have recently been changed

### CODE REVIEW

File: frontend/src/APIClients/types/TestSessionClientTypes.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export interface TestSessionMetadata {
+  /** the unique identifier for the test session */
+  id: string;
+  /** on this date, the test becomes available to students */
+  startDate: Date;
+  /** after this date, the test is no longer available to students */
+  endDate: Date;
+  /** notes inputted by teacher to show students prior to commencing the test */
+  notes?: string;
+}
```
Comment (joyce-shi) [inline frontend/src/APIClients/types/TestSessionClientTypes.ts:10]: i put this in `TestSessionClientTypes` cuz its really the interface for the response without certain fields. not sure if we want to move this somewhere else or rename `TestSessionMetadata` to `TestSessionMetadataResponse` or something

File: frontend/src/components/pages/student/AssessmentSummaryPage.tsx [hunk 2]
```diff
@@ -29,29 +29,41 @@ const AssessmentSummaryPage = (): React.ReactElement => {
 
   return (
     <>
-      {test && (
+      {test && testSession && (
```
Comment (joyce-shi) [inline frontend/src/components/pages/student/AssessmentSummaryPage.tsx:38]: I don't know how I feel about this b/c technically this page would never show if these fields were null... (b/c that case is handled in `StudentDashboard` ), but idk how to make this better lol

**Comment by joyce-shi** [comment]
@JennaXiao feel free to take a look & leave a review, since some of the `StudentContext` changes feed into the student assessment experience you have ownership over. (:

File: backend/services/implementations/testSessionService.ts [hunk 3]
```diff
@@ -104,19 +108,21 @@ class TestSessionService implements ITestSessionService {
     accessCode: string,
   ): Promise<TestSessionResponseDTO> {
     let testSessionDtos: Array<TestSessionResponseDTO> = [];
-
+    const currentDate = new Date();
     try {
       const testSessions: Array<TestSession> = await MgTestSession.find({
         accessCode: { $eq: accessCode },
+        startDate: { $lte: currentDate },
+        endDate: { $gte: currentDate },
```
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts:121]: side note, we should prob make explicit to user somewhere that dates are inclusive (if its not in designs)

File: backend/services/implementations/testSessionService.ts [hunk 4]
```diff
@@ -62,8 +62,11 @@ class TestSessionService implements ITestSessionService {
       teacherDTO = await this.userService.getUserById(testSession.teacher);
       schoolDTO = await this.schoolService.getSchoolById(testSession.school);
 
-      newTestSession = await MgTestSession.create(testSession);
+      if (testSession.startDate > testSession.endDate) {
```
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts]: should this be something controlled by the FE? another check would maybe for if endDate < currentDate

if FE handles, then we should update docs to mention this guarantee

File: frontend/src/components/pages/student/AssessmentSummaryPage.tsx [hunk 5]
```diff
@@ -11,14 +11,15 @@ import {
 
 import { JUMP_MATH_LOGO } from "../../../assets/images";
 import { HOME_PAGE } from "../../../constants/Routes";
-import { assessmentMetadata } from "../../../constants/StudentAssessmentConstants";
 import AuthContext from "../../../contexts/AuthContext";
 import StudentContext from "../../../contexts/StudentContext";
+import { getReadableDateTime } from "../../../utils/GeneralUtils";
 import AssessmentInfo from "../../assessments/student-experience/AssessmentInfo";
 import AssessmentRules from "../../assessments/student-experience/AssessmentRules";
+import ErrorState from "../../common/ErrorState";
```
Comment (carissa-tang) [inline frontend/src/components/pages/student/AssessmentSummaryPage.tsx]:
```suggestion
```

File: frontend/src/components/pages/student/AssessmentSummaryPage.tsx (hunk 2 above)
Comment (carissa-tang) [reply frontend/src/components/pages/student/AssessmentSummaryPage.tsx:38]: hmm its fine lol. here's a [thread](https://stackoverflow.com/questions/67785257/best-practice-when-asserting-non-null-on-a-value-coming-from-react-usecontext-o) of ideas but i feel like it's kinda extra

File: frontend/src/APIClients/types/TestSessionClientTypes.ts (hunk 1 above)
Comment (carissa-tang) [reply frontend/src/APIClients/types/TestSessionClientTypes.ts:10]: i think this is fine

**Review by carissa-tang — COMMENTED** [review]
can we put a min width to these icons LOL
image

File: backend/services/implementations/testSessionService.ts (hunk 3 above)
Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts:121]: yup. I think we can ask design to add it in the copy when creating the form

File: backend/services/implementations/testSessionService.ts (hunk 4 above)
Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts]: i think in general its best practice to do validation on both the frontend and backend? i don't think super critical to have for our use case, but also why not 🤷‍♀️

https://www.quora.com/Should-I-do-input-validation-on-the-front-end-or-back-end
https://stackoverflow.com/questions/37492271/should-i-handle-html-form-validation-on-front-or-backend

but good point for the `end < current`, I will add that!

Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts:65]: wait `end > current` right? :D

**Review by carissa-tang — APPROVED** [review]
lgtm!!

**Pull request merged**